store/tikv/mocktikv/aggregate.go

```diff
@@ -0,0 +1,302 @@
+		}
+		bufLen += len(b)
+		row = append(row, b)
+	}
+	buf := make([]byte, 0, bufLen)
+	for _, col := range row {
+		buf = append(buf, col...)
+	}
+	return buf, row, nil
+}
+
+// aggregate updates aggregate functions with row.
+func (e *hashAggExec) aggregate(value [][]byte) error {
+	err := e.evalCtx.decodeRelatedColumnVals(e.relatedColOffsets, value, e.row)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Get group key.
+	gk, gbyKeyRow, err := e.getGroupKey()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if _, ok := e.groups[string(gk)]; !ok {
+		e.groups[string(gk)] = struct{}{}
+		e.groupKeys = append(e.groupKeys, gk)
+		e.groupKeyRows = append(e.groupKeyRows, gbyKeyRow)
+	}
+	// Update aggregate expressions.
+	aggCtxs := e.getContexts(gk)
+	for i, agg := range e.aggExprs {
+		err = agg.Update(aggCtxs[i], e.evalCtx.sc, e.row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *hashAggExec) getContexts(groupKey []byte) []*aggregation.AggEvaluateContext {
+	groupKeyString := string(groupKey)
+	aggCtxs, ok := e.aggCtxsMap[groupKeyString]
+	if !ok {
+		aggCtxs = make([]*aggregation.AggEvaluateContext, 0, len(e.aggExprs))
+		for _, agg := range e.aggExprs {
+			aggCtxs = append(aggCtxs, agg.CreateContext())
+		}
+		e.aggCtxsMap[groupKeyString] = aggCtxs
+	}
+	return aggCtxs
+}
+
+type streamAggExec struct {
+	evalCtx           *evalContext
+	aggExprs          []aggregation.Aggregation
+	aggCtxs           []*aggregation.AggEvaluateContext
+	groupByExprs      []expression.Expression
+	relatedColOffsets []int
+	row               types.DatumRow
+	tmpGroupByRow     types.DatumRow
+	currGroupByRow    types.DatumRow
+	nextGroupByRow    types.DatumRow
+	currGroupByValues [][]byte
+	executed          bool
+	hasData           bool
+
+	src executor
+}
+
+func (e *streamAggExec) SetSrcExec(exec executor) {
+	e.src = exec
+}
+
+func (e *streamAggExec) getPartialResult() ([][]byte, error) {
+	value := make([][]byte, 0, len(e.groupByExprs)+2*len(e.aggExprs))
+	for i, agg := range e.aggExprs {
+		partialResults := agg.GetPartialResult(e.aggCtxs[i])
+		for _, result := range partialResults {
+			data, err := codec.EncodeValue(nil, result)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			value = append(value, data)
+		}
+		// Clear the aggregate context.
+		e.aggCtxs[i] = agg.CreateContext()
+	}
+	e.currGroupByValues = e.currGroupByValues[:0]
+	for _, d := range e.currGroupByRow {
+		buf, err := codec.EncodeValue(nil, d)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.currGroupByValues = append(e.currGroupByValues, buf)
+	}
+	e.currGroupByRow = e.nextGroupByRow.Copy()
+	return append(value, e.currGroupByValues...), nil
+}
+
+func (e *streamAggExec) meetNewGroup(row [][]byte) (bool, error) {
+	if len(e.groupByExprs) == 0 {
+		return false, nil
+	}
+
+	e.tmpGroupByRow = e.tmpGroupByRow[:0]
+	matched, firstGroup := true, false
+	if e.nextGroupByRow == nil {
```

Comment:
Should we use `currGroupByRow` to check ?

Reply:
I think both are OK.